### PR TITLE
658 contains pattern action

### PR DESCRIPTION
**Pull request by Mefpef**
Type pull request title with short description (#658 )

Contains pattern action plugin #658 

done:
plugin
tests
register plugin

----

**Please** following this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Give the same name to the pull request as the name of ticket 
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Test your code throughly before PR.
 
If your pull request is about ~20 lines of code add license information in your code in form of comment. i.e. # License: MIT, Author:  Your name.

To make clear that you license your contribution under the MIT you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [MIT license](https://opensource.org/licenses/MIT)

PRs without ticked MIT license will not be accepted - Sorry.

### CODE REVIEW

File: tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+from tracardi.service.plugin.domain.result import Result
+
+
+class Config(PluginConfig):
+    field: str
+    pattern: str
+
+    @validator("field")
+    def validate_prefix(cls, value):
+        if value == "":
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py:18]: "if not value" is gonna cover more cases

File: tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+from tracardi.service.plugin.domain.config import PluginConfig
+from tracardi.service.plugin.domain.result import Result
+
+
+class Config(PluginConfig):
+    field: str
+    pattern: str
+
+    @validator("field")
+    def validate_prefix(cls, value):
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py:17]: validate_field

File: tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        if value == "":
+            raise ValueError("Field cannot be empty")
+        return value
+
+    @validator("pattern")
+    def validate_field(cls, value):
+        if value == "":
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py:24]: same as above

File: tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    config: Config
+
+    async def set_up(self, init):
+        self.config = validate(init)
+
+    async def run(self, payload: dict, in_edge=None):
+        dot = self._get_dot_accessor(payload)
+        value = dot[self.config.field]
+
+        if not isinstance(value, str) or isinstance(value, list):
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py:53]: if not (isintance(value, str) or isintance(value, list)). For now, List will raise an error that is unwanted.

File: tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            raise WrongFieldTypeError(f"Given field must be an array or string type. {type(value)} given")
+
+
+        founded_patterns = []
+        if self.config.pattern == "all":
+            for k, v in patterns.items():
+                if re.findall(patterns[k], value):
+                    founded_patterns.append(k)
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py:61]: found, founded stands for "ufundowany"

File: tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+            return Result(port='true', value={"found": founded_patterns})
+
+        if self.config.pattern == "email":
+            if re.match(patterns["email"], value):
+                return Result(port="true", value={"found": "email"})
+
+        if self.config.pattern == "url":
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py:68]: use elif, slight performance increase

File: tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+        if self.config.pattern == "date":
+            if re.match(patterns["date"], value):
+                return Result(port="true", value={"found": "date"})
+
+        if self.config.pattern == "ip":
+            if re.match(patterns["ip"], value):
+                return Result(port="true", value={"found": "ip"})
+
+        return Result(port="false", value="Didn't find any patterns")
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py:80]: do not return string as value. That is going to raise an error in the further execution. Return something like value={"found": None} or {"message": "Did not find any pattern"}.

File: tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+            desc='Checks if field contains defined pattern.',
+            icon='question',
+            group=["Flow control"],
+            documentation=Documentation(
+                inputs={
+                    "payload": PortDoc(desc="This port takes payload object.")
+                },
+                outputs={"true": PortDoc(desc="This port returns payload if field contains defined pattern."),
+                         "false":
+                             PortDoc(desc="This port returns payload if field doesnt contain defined pattern.")}
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py:149]: doesn't contain

File: tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        metadata=MetaData(
+            name='Contains pattern',
+            desc='Checks if field contains defined pattern.',
+            icon='question',
+            group=["Flow control"],
+            documentation=Documentation(
+                inputs={
+                    "payload": PortDoc(desc="This port takes payload object.")
+                },
+                outputs={"true": PortDoc(desc="This port returns payload if field contains defined pattern."),
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py:147]: these outputs does not return payload. They return new objects.

File: tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+    async def run(self, payload: dict, in_edge=None):
+        dot = self._get_dot_accessor(payload)
+        value = dot[self.config.field]
+
+        if not isinstance(value, str) or isinstance(value, list):
+            raise WrongFieldTypeError(f"Given field must be an array or string type. {type(value)} given")
+
+
+        founded_patterns = []
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py:57]: should be declared below "if" statement in the next line.

**Review by DawidekZagajnik — CHANGES_REQUESTED** [review]

File: tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py (hunk 1 above)
Comment (DawidekZagajnik) [reply tracardi/process_engine/action/v1/operations/contains_pattern/plugin.py:18]: one more thing: make this field an Enum, so it checks if it is one of email, ip, date, etc. even if the plugin is configured using advanced config.

**Review by atompie — CHANGES_REQUESTED** [review]
Could you please make the changes suggested by David. Thanks.

**Pull request merged**